dnsdist: Expose trailing data

gibson042 opened the pull request:
### Short description
Expose data following a DNS message to Lua for both reading and writing.
Fixes #6846.

Incorporates #6897 ([diff from that base](https://github.com/gibson042/pdns/compare/2018-08-trailing-data-tests...2018-08-trailing-data)).

### Checklist
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [x] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

rgacogne commented on regression-tests.dnsdist/test_Trailing.py:
This line looks like a c/p mistake.

gibson042 replied:
No, the idea is that queries for replaced.trailing.tests.powerdns.com will be subject first to `replaceTrailingData` (which sets the data to ASCII "ABC") and then to `reportTrailingData` (which uses trailing data to spoof a response), allowing us to verify that the first succeeded.

rgacogne replied:
ah, right, I missed that! Thanks for the explanation!

rgacogne commented on pdns/dnsdist-lua-bindings-dnsquestion.cc:
```diff
@@ -54,6 +54,27 @@ void setupLuaBindingsDNSQuestion()
+      dq.len = length + data.size();
+      uint8_t* tail = message + length - 1;
+      for(const auto& pair : data) {
+        *(tail + pair.first) = pair.second;
```
I'm probably paranoid but I'd prefer the use of a counter instead of using the indexes passed by `Lua`.

gibson042 replied:
I was also wondering about that... but is the order of data from Lua tables guaranteed? And if not, should I just guard the write like e.g.
```cpp
int size = data.size();
for(const auto& pair : data) {
  if(pair.first < 1 || pair.first > size) {
    /* reject sparse input */
    return false;
  }
  *(tail + pair.first) = pair.second;
}
```

rgacogne replied:
To be fair if the order is not right I'd consider it a bug in the wrapper. Still I'd like some sort of check, yes, but we need to be careful because we already increased `dq.len` at that point.

gibson042 replied:
So what kind of protection would you like to see?

gibson042 replied:
@rgacogne would it make more sense to follow what #6068 is doing with EDNS option data and expose the trailing data as a string rather than a uint8 list? It would work as long as 0x00 is allowed and strings are sufficiently manipulable in Lua, and would resolve the ordering concern.

rgacogne replied:
Sorry it took me so long to answer this comment :-/
Yes, we have been using binary-safe strings in dnsdist for this kind of things so far, so IMHO it would be a better choice. Again, thanks a lot for contributing.

gibson042 commented:
@rgacogne I updated this PR to expose the data as a string rather than an array of bytes, and added tests for non-ASCII octets in the trailing data (both inbound and outbound). This should be ready for final review.

rgacogne commented on pdns/dnsdist-lua-bindings-dnsquestion.cc:
```diff
@@ -63,6 +63,27 @@ void setupLuaBindingsDNSQuestion()
+      char* message = reinterpret_cast<char*>(dq.dh);
+      const uint16_t messageLen = getDNSPacketLength(message, dq.len);
+      const uint16_t tailLen = tail.size();
+      if(messageLen + tailLen > dq.size) {
```
Not a big deal but it looks like `messageLen` + `tailLen` could overflow before being promoted. Perhaps something like:
```suggestion
      if(tailLen > (dq.size - messageLen)) {
```

gibson042 replied:
Nice catch! Updated.

gibson042 commented on regression-tests.dnsdist/dnsdisttests.py:
```diff
@@ -142,42 +142,54 @@ def _ResponderIncrementCounter(cls):
+            except dns.message.TrailingJunk as e:
+                if trailingDataResponse is False:
+                    raise
+                print("UDP query with trailing data, synthesizing response")
```
The existing structure was already a bit confusing, and I was trying to preserve it as much as possible (which unfortunately didn't help). But the idea is to extend the existing `ignoreTrailing` parameter into `trailingDataResponse` as follows:
* `False`: reject queries with trailing data (_preexisting behavior_)
* `True`: ignore trailing data after queries (_preexisting behavior_)
* anything else: interpret as RCODE to be sent in response to queries with trailing data (_new behavior, used by `TestTrailingDataToBackend` in regression-tests.dnsdist/test_Trailing.py_)
We will never pass `synthesize=True` to `_getResponse`, because `forceRcode` is only set to `trailingDataResponse` when dns.message.from_wire raises dns.message.TrailingJunk, which cannot happen when it is ignoring trailing data.

I have added comments and safety-enforcing code.

gibson042 commented:
@rgacogne Why was dnsdist 1.3.3 released including [DNSQuestion:getEDNSOptions()](https://github.com/PowerDNS/pdns/commit/ef8640ddc0fcff7c84b955f2ec2110939140ca2e) but not these additions?

rgacogne commented:
> Why was dnsdist 1.3.3 released including DNSQuestion:getEDNSOptions() but not these additions?

The main reason is that we are having a hard time deciding whether we want to make it possible to use trailing data within dnsdist. I personally don't like the idea too much, but if people are going to do it anyway I'd prefer the feature to be merged and be done with it.

gibson042 commented:
> if people are going to do it anyway I'd prefer the feature to be merged and be done with it.

I really wish you would do that. Remember that this feature also provides a means of [protecting downstream servers _from_ trailing data](https://github.com/PowerDNS/pdns/pull/6967/files#diff-f25095fe509924812e11dd26ab8ea6b6R179):

```lua
function removeTrailingData(dq)
    local success = dq:setTrailingData("")
    if not success then
        return DNSAction.ServFail, ""
    end
    return DNSAction.None, ""
end
```

rgacogne commented:
Sorry it took so long, thanks a lot for this pull request!
